src/librustc_trans/mir/block.rs

```diff
@@ -334,14 +334,12 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
                         let len = self.trans_operand(&mut bcx, len).immediate();
                         let index = self.trans_operand(&mut bcx, index).immediate();
 
-                        let const_err = common::const_to_opt_uint(len).and_then(|len| {
-                            common::const_to_opt_uint(index).map(|index| {
-                                ErrKind::IndexOutOfBounds {
-                                    len: len,
-                                    index: index
-                                }
-                            })
-                        });
+                        let const_err = common::const_to_opt_u128(len, false)
+                            .and_then(|len| common::const_to_opt_u128(index, false)
+                                .map(|index| ErrKind::IndexOutOfBounds {
```

Comment:
The index is assured by typecheck to be `usize` and therefore can't exceed the range of 64 bits, can it? Would it be an idea to re-add `const_to_opt_uint` for cases like this and the one you mentioned above?